src/sonardevices/OculusSonar.cpp

```diff
@@ -0,0 +1,555 @@
+        }
+        OculusMessages::OculusMessageHeader *omh = (OculusMessages::OculusMessageHeader *)rxBuff;
+
+        // Check if data is valid via oculus ID
+        if (omh->oculusId != OCULUS_CHECK_ID)
+        {
+            // Data invalid, flush buffer
+            rxBuffPos = 0;
+            continue;
+        }
+
+        currPktSize += omh->payloadSize;
+
+        // Check if payload is also in the buffer and process message
+        if (rxBuffPos < currPktSize){
+            continue;
+        }
+
+        // Check which message type was received
+        switch (omh->msgId){
+            case OculusMessages::OculusMessageType::messageSimplePingResult:
+                processSimplePingResult((OculusMessages::OculusSimplePingResult *)omh);
+                break;
+            case OculusMessages::OculusMessageType::messageUserConfig:
+                printf("OculusSonar: Received user config message\n");
+                break;
+            case OculusMessages::OculusMessageType::messageDummy:
+                printf("OculusSonar: Received dummy message\n");
+                break;
+            default:
+                printf("OculusSonar: Received unknown message\n");
+                break;
+        }
+
+        // If there is additional data in buffer shift it to the front
+        memmove(rxBuff, &rxBuff[currPktSize], rxBuffPos - currPktSize);
+        rxBuffPos -= currPktSize;
+    
+    }
+
+    // Free memory
+    free(rxBuff);
+    rxBuff = nullptr;
+}
+
+void OculusSonar::processSimplePingResult(OculusMessages::OculusSimplePingResult *ospr){
+    uint8_t *startAddress = (uint8_t *)ospr;
+
+    uint32_t imageSize;
+    uint32_t imageOffset;
+    uint16_t beams;
+    uint16_t ranges;
+
+    uint16_t version = ospr->fireMessage.head.msgVersion;
+    
+    OculusMessages::OculusSimplePingResult2 *ospr2 = (OculusMessages::OculusSimplePingResult2 *)ospr;
+
+    // Check message version
+    switch (version){
+        case 2:
+            imageSize = ospr2->imageSize;
+            imageOffset = ospr2->imageOffset;
+            beams = ospr2->nBeams;
+            ranges = ospr2->nRanges;
+            rangeResolution = ospr2->rangeResolution;
+            break;
+        default:
+            imageSize = ospr->imageSize;
+            imageOffset = ospr->imageOffset;
+            beams = ospr->nBeams;
+            ranges = ospr->nRanges;
+            rangeResolution = ospr->rangeResolution;
+            break;
+    }
+    // Remember the beams and range bin count
+    this->beams = beams;
+    this->rangeBinCount = ranges;
+
+    // Check if image size is correct
+    if (ospr->fireMessage.head.payloadSize + sizeof(OculusMessages::OculusMessageHeader) == imageOffset + imageSize){
+        switch (version){
+            case 2:{
+                this->operatingFrequency = ospr2->frequency;
+                this->temperature = ospr2->temperature;
+                this->pressure = ospr2->pressure;
+                break;
```

Comment:
@EulersBoiler Missing handling of heading pitch and roll from onboard IMU of certain sonar models. This breaks support for a whole range of oculus sonars.